### PR TITLE
ENH: Add itkPolyDataToMeshFilter

### PR DESCRIPTION
**Pull request by tbirdso**
Adds `itk::PolyDataToMeshFilter` with testing and Jupyter Notebook example.

### CODE REVIEW

**Comment by tbirdso** [comment]
Currently viewing the following error when trying to execute `filter->Update()` in `itkPolyDataToMeshFilterTest`:
> 6: Trying filter->Update()
6:
6: itk::ExceptionObject (000000C0F00FF8A8)
6: Location: "unknown"
6: File: C:\repos\ITK\Modules\Core\Common\include\itkPointSet.hxx
6: Line: 361
6: Description: ITK ERROR: Mesh(000002153C4B9190): itk::PointSet::CopyInformation() cannot cast class itk::DataObject const * __ptr64 to class itk::PointSet<double,3,class itk::DefaultStaticMeshTraits<double,3,3,float,float,double> > * __ptr64
6:

I've tried using breakpoints in `GenerateData()` and commenting out existing code, it appears that `GenerateData` is not getting called. Current guess is that there is an underlying component of `ProcessObject` that I'm misunderstanding. @thewtex @dzenanz do you have any insight into this issue?

File: include/itkPolyDataToMeshFilter.h [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+     *
+     * \brief Convert PolyData to a Mesh
+     *
+     * Convert an itk::PolyData to an itk::PointSet or itk::Mesh
+     *
+     * \ingroup MeshToPolyData
+     *
+     */
+    template< typename TInputMesh >
+    class PolyDataToMeshFilter : public ProcessObject
```
Comment (dzenanz) [inline include/itkPolyDataToMeshFilter.h:39]: Could you inherit from [MeshToMeshFilter](https://itk.org/Doxygen/html/classitk_1_1MeshToMeshFilter.html)?

File: include/itkPolyDataToMeshFilter.hxx [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+    }
+
+
+    template< typename TInputMesh >
+    void
+        PolyDataToMeshFilter< TInputMesh >
+        ::GenerateData()
+    {
+        const InputMeshType* inputMesh = this->GetInput();
+        PolyDataType* outputPolyData = this->GetOutput();
```
Comment (dzenanz) [inline include/itkPolyDataToMeshFilter.hxx]: Is calling `AllocateOutputs()` required?

Comment (tbirdso) [reply include/itkPolyDataToMeshFilter.hxx]: Outputs should be allocated in the constructor with `MakeOutput`:
> template< typename TInputPolyData >
    PolyDataToMeshFilter< TInputPolyData >
        ::PolyDataToMeshFilter()
    {
        // Modify superclass default values, can be overridden by subclasses
        this->SetNumberOfRequiredInputs(1);
        typename MeshType::Pointer output = static_cast<MeshType*>(this->MakeOutput(0).GetPointer());
        this->ProcessObject::SetNumberOfRequiredOutputs(1);
        this->ProcessObject::SetNthOutput(0, output.GetPointer());
    }

It looks like `AllocateOutputs` is defined in `itkImageTransformer`, which inherits from `ProcessObject` but is not in the inheritance hierarchy of this filter.

File: include/itkPolyDataToMeshFilter.h (hunk 1 above)
Comment (tbirdso) [reply include/itkPolyDataToMeshFilter.h:39]: In this case the input type is not an `itk::Mesh` so it is probably better to not inherit from `MeshToMeshFilter`? Ideally we can mirror the format of `itkMeshToPolyDataFilter` which also inherits from `ProcessObject`.

**Comment by thewtex** [comment]
@tbirdso the default behavior of `ProcessObject::GenerateOutputInformation` calls `CopyInformation` from the first input to the output.

https://github.com/InsightSoftwareConsortium/ITK/blob/6dfbc9f0314b5fd64d4b2ae392ae3cc6dda17677/Modules/Core/Common/src/itkProcessObject.cxx#L1803-L1822

You will need to override `GenerateOutputInformation`, use `itk::Mesh::CopyInformation` as a reference:

https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Mesh/include/itkMesh.hxx

(it currently calls `PointSet::CopyInformation`, but substitute what is done there, etc.).

**Comment by tbirdso** [comment]
Thanks @thewtex! Overriding `GenerateOutputInformation` so that it no longer tries to copy metadata to the `itk::Mesh` output from the `itk::PolyData` input solved the issue. I've referenced the [ITK Software Guide](https://itk.org/ITKSoftwareGuide/html/Book1/ITKSoftwareGuide-Book1ch8.html#x54-1290008) for an explanation of where this fits in the ITK pipeline.

The default metadata of `itk::PointSet` [set in its constructor](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Core/Common/include/itkPointSet.hxx#L311-L318) is to allow exactly one region for representing unstructured data. If I'm understanding correctly, can we simply do nothing in the `GenerateOutputInformation` override and leave the output mesh generated from the input polydata with its default metadata?

File: include/itkPolyDataToMeshFilter.hxx [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+#include "itkPolyDataToMeshFilter.h"
+
+#include "itkVertexCell.h"
+#include "itkLineCell.h"
+#include "itkMesh.h"
+#include "itkTriangleCell.h"
+#include "itkQuadrilateralCell.h"
+#include "itkPolygonCell.h"
+ //#include "itkTetrahedronCell.h"
+ //#include "itkHexahedronCell.h"
```
Comment (thewtex) [inline include/itkPolyDataToMeshFilter.hxx:30]: Remove

File: include/itkPolyDataToMeshFilter.hxx [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+
+namespace itk
+{
+
+    template< typename TInputPolyData >
+    PolyDataToMeshFilter< TInputPolyData >
```
Comment (thewtex) [inline include/itkPolyDataToMeshFilter.hxx:36]: Please add the *ITK/.clang-format* file and run `clang-format-8` on the code (the binary is pull down in the ITK build tree).. I

File: include/itkPolyDataToMeshFilter.hxx [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+        const CellContainerType* inputVertices = inputPolyData->GetVertices();
+        inputCellItr = inputVertices->Begin();
+        inputCellEnd = inputVertices->End();
+
+        while (inputCellItr != inputCellEnd) {
+            auto numPoints = inputCellItr.Value();
+            ++inputCellItr;
+
+            // Verify vertex contains exactly one point ID
+            if (numPoints == VertexCellType::NumberOfPoints) {
```
Comment (thewtex) [inline include/itkPolyDataToMeshFilter.hxx:223]: We could use `itkAssertInDebugAndIgnoreInReleaseMacro` for these checks to notify the developer of a logical error in `Debug` mode but avoid performance penalties in `Release`.

**Review by thewtex — CHANGES_REQUESTED** [review]
@tbirdso very nice!! 🏅 🏅 

A few minor style issues noted inline.

**Review by thewtex — APPROVED** [review]
🚀

**Pull request merged**